server/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -890,23 +893,31 @@ public void createShard(
         RecoveryState recoveryState = indexService.createRecoveryState(shardRouting, targetNode, sourceNode);
         IndexShard indexShard = indexService.createShard(shardRouting, globalCheckpointSyncer, retentionLeaseSyncer);
         indexShard.addShardFailureCallback(onShardFailure);
-        indexShard.startRecovery(recoveryState, recoveryTargetService, recoveryListener, repositoriesService, (mapping, listener) -> {
-            assert recoveryState.getRecoverySource().getType() == RecoverySource.Type.LOCAL_SHARDS
-                : "mapping update consumer only required by local shards recovery";
-            AcknowledgedRequest<PutMappingRequest> putMappingRequestAcknowledgedRequest = new PutMappingRequest().setConcreteIndex(
-                shardRouting.index()
-            )
-                .setConcreteIndex(shardRouting.index()) // concrete index - no name clash, it uses uuid
-                .source(mapping.source().string(), XContentType.JSON);
-            // concrete index - no name clash, it uses uuid
-            client.execute(
-                featureService.clusterHasFeature(clusterService.state(), SUPPORTS_AUTO_PUT)
-                    ? TransportAutoPutMappingAction.TYPE
-                    : TransportPutMappingAction.TYPE,
-                putMappingRequestAcknowledgedRequest.ackTimeout(TimeValue.MAX_VALUE).masterNodeTimeout(TimeValue.MAX_VALUE),
-                new RefCountAwareThreadedActionListener<>(threadPool.generic(), listener.map(ignored -> null))
-            );
-        }, this, clusterStateVersion);
+        indexShard.startRecovery(
+            recoveryState,
+            recoveryTargetService,
+            postRecoveryMerger.maybeMergeAfterRecovery(shardRouting.shardId(), recoveryListener),
```

Comment:
```suggestion
shardRouting.isPromotableToPrimary() ? postRecoveryMerger.maybeMergeAfterRecovery(shardRouting.shardId(), recoveryListener) : recoveryListener,
```

Reply:
I don't think we need this - calling `forceMerge()` on the engine under an unpromotable shard is a no-op anyway.

Reply:
Looks an easy check to avoid enqueing unnecessary work to me :)

Reply:
actually yeah I'll do that, this extra machinery isn't totally free after all